### PR TITLE
Conditionally add subscription payment gateway features

### LINKED ISSUE #3819: Incorrect subscription support flags added when WCS is inactive
### Describe the bug
Several subscription support flags are added to the WCPay payment gateway when WooCommerce Subscription is inactive and WCPay Subscriptions is enabled.

Ref: https://github.com/Automattic/woocommerce-payments/blob/a50c1079ea885dcb9bc2ae4af88175d678365c94/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php#L95-L107

Some of these support flags are only valid when WCS is enabled and **will not work** with WCPay Subscriptions, e.g. `subscription_amount_changes`.

### Actual behavior
The WCPay payment gateway adds the full set of subscription support flags regardless of whether WCS is active.

### Expected behavior
The WCPay payment gateway adds a limited set of subscriptions support flags when WCPay Subscriptions is active and WCS is not active and adds the full set of subscription support flags when WCS is active.

### Additional context
This was discovered when attempting to add WCPay Subscriptions compatibility to AutomateWoo in 1092-gh-woocommerce/automatewoo.

### PR DESCRIPTION
**Pull request by aprea**
Fixes #3819 

#### Changes proposed in this Pull Request

This PR reduces the [supported payment gateway features](https://woocommerce.github.io/code-reference/classes/WC-Payment-Gateway.html#property_supports) when the WooCommerce Subscriptions plugin is not active.

In particular, the `subscription_amount_changes` and `subscription_date_changes` are only included when WCS is active (as I believe these features do not work with WCPay subscriptions).

And the `gateway_scheduled_payments` feature is included when WCS is not active. I'm not 100% sure about this one but I saw an existing instance of this here → https://github.com/Automattic/woocommerce-payments/blob/32714ed/includes/subscriptions/class-wc-payments-subscription-service.php#L81

All existing subscriptions functionality that is expected to work when WCS is disabled should continue to work with these changes.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

**Comment by mattallan** [comment]
@aprea the changes in this PR look good to me!

> And the gateway_scheduled_payments feature is included when WCS is not active.

Yeah this is good :+1:

I did have some questions around the difference between `$gateway->supports( $feature )` and `$subscription->payment_method_supports( $feature )` when WC Subscriptions plugin is disabled.

When calling `$subscription->payment_method_supports( $feature )` the following features are allowed:
 - `'gateway_scheduled_payments'`
 - `'subscriptions'`
 - `'subscription_suspension'`
 - `'subscription_reactivation'`
 - `'subscription_cancellation'`

However, `$gateway->supports( $feature )` allows:
 - `'gateway_scheduled_payments'`
 - `'subscriptions'`
 - `'subscription_suspension'`
 - `'subscription_reactivation'`
 - `'subscription_cancellation'`
 - `'subscription_payment_method_change_admin'`
 - `'subscription_payment_method_change_customer'`
 - `'subscription_payment_method_change'`
 - `'multiple_subscriptions'`

cc @james-allan I'm curious to get your thoughts on the above regarding some features returning true when calling `$gateway->supports()` and false when calling `subscription->payment_method_supports()` when WC Subscriptions is disabled.

I'm guessing it doesn't matter because there isn't any code where `payment_method_supports( 'subscription_payment_method_change_admin' )` is called so from that perspective this PR looks good to me :)

**Review by mattallan — APPROVED** [review]
I've just a left a comment on this and pinged James to see if he has any opinions on this as well (considering this is a fairly large change 😅)

I've approved the PR as it looks good to me 💯

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php [hunk 1]
```diff
@@ -92,19 +92,30 @@ public function maybe_init_subscriptions() {
...
+		if ( $this->is_subscriptions_plugin_active() ) {
+			$payment_gateway_features = array_merge(
+				$payment_gateway_features,
+				[
+					'subscription_amount_changes',
+					'subscription_date_changes',
+				]
+			);
+		} else {
+			$payment_gateway_features[] = 'gateway_scheduled_payments';
```
[the remark this answers is not in the document]
Comment (aprea) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:128]: `gateway_scheduled_payments` **should not** be added when WCS is active so I don't think I can add this to the initial array.

I have added some comments to clarify this.

**Comment by james-allan** [comment]
> I'm curious to get your thoughts on the above regarding some features returning true when calling $gateway->supports() and false when calling subscription->payment_method_supports() when WC Subscriptions is disabled.

yeah I'm not sure why I only included a subset to begin with. IIRC I just added the ones I was interested in testing at the time and figured we'd add more features as they popped up.  

If `$subscription->payment_method_supports( 'subscription_payment_method_change_customer' )` isn't currently supported, then I'm not sure how customers can change the payment method at the moment. 

I think it would make sense that they are more consistent.

**Comment by mattallan** [comment]
> then I'm not sure how customers can change the payment method at the moment.

This is because when showing the "Change Payment" option on the My Account > View Subscription page, we're using `$subscription->can_be_updated_to( 'new-payment-method' )` which calls:

 - `WC_Subscriptions_Core_Payment_Gateways::one_gateway_supports( 'subscription_payment_method_change_customer' )`, which calls
 - `$gateway->supports( 'subscription_payment_method_change_customer' )`

So `$subscription->payment_method_supports( 'subscription_payment_method_change_customer' )` is never called :)

**Comment by aprea** [comment]
Hi @james-allan & @mattallan, I have mirrored the supported features to `includes/subscriptions/class-wc-payments-subscription-service.php` (i.e. in the `WC_Payments_Subscription_Service` class) in e24d18f580f4a95d25b4eb65fa0c821a50344f9e to keep things consistent.

Does this look like the right approach?

**Comment by mattallan** [comment]
Apologies @aprea for not getting back to this yesterday!

> Does this look like the right approach?

Yep! 🙌 At first I was thinking "if they're mirrored why do we still need to have this second supports array?" but then I realised it's possible where WC Subscription extension is active and we're checking `$subscription->payment_method_supports( 'x_feature' )` on a Stripe billing subscription so we still need to make sure some features are limited.

The latest changes look good! 💯

**Pull request merged**